Add Notifications Drawer and Toast Notifications List

jeff-phillips-18 opened the pull request:
Adding a notifications drawer to provide a mechanism to show users events/tasks in the UI. Built using the Angular Patternfly directive, the drawer shows on the right side of the window above the main content. It can be opened/closed via a notifications indicator on the top nav bar.

The Toast Notifications list provides a place to show popup notifications to the user. The notifications are shown for 8 seconds for normal notifications but remain until closed for error notifications.

This PR does not cover retrieving  the notifications from the server or persisting any notification states.

Notification Drawer: 
![image](https://cloud.githubusercontent.com/assets/11633780/18181476/47228f64-7059-11e6-809c-37b44a3f2d42.png)

Toast Notifications:
![image](https://cloud.githubusercontent.com/assets/11633780/18180847/a4f17842-7056-11e6-8395-dfba333da8fe.png)


References for Patterns on Patternfly:
-------------------------------------------
> * https://www.patternfly.org/pattern-library/communication/notification-drawer/#/api
> * https://www.patternfly.org/pattern-library/communication/toast-notifications/#/api
> * http://angular-patternfly.rhcloud.com/#/api/patternfly.notification.directive:pfToastNotificationList


cc @skateman @serenamarie125

serenamarie125 commented:
Looks great @jeff-phillips-18 .  For those who don't know we have already done a round of usability testing of this PatternFly component at RH Summit

dclarizio commented:
@himdel @skateman please review and let me know if this is ready to go so you can move forward on the notifications work.

himdel commented on app/assets/javascripts/miq_debug.js:
```diff
@@ -71,7 +71,23 @@ $(function() {
 });
 
-angular.module('miq.debug', [])
+function sanitize (data) {
```
We should use a better name here now that it's a global.. `sanitizeError` maybe?

himdel commented on app/assets/javascripts/miq_debug.js:
```diff
@@ -21,7 +21,7 @@ $(document).bind('keyup', 'ctrl+shift+x', miqSparkleOff);
   var orig = {
     error: window.console.error,
-    warn: window.console.warn,
+    warn: window.console.warn
```
Uh, can you set your editor to *not* remove trailing commas please?

himdel commented on app/views/layouts/_header.html.haml:
```diff
@@ -32,6 +35,9 @@
   = render :partial => "layouts/lightbox_panel"
 
+:javascript
+  miq_bootstrap('.navbar.navbar-pf-vertical');
```
I wonder if we should group this notification related functionality in a separate ng module, something like `miq.notifications` perhaps? That way we can just instantiate that one here..

himdel commented on app/views/layouts/_notifications_drawer.html.haml:
```diff
@@ -0,0 +1,17 @@
+       "custom-scope" => "vm.customScope"}
+
+:javascript
+  miq_bootstrap('#miq-notifications-drawer-container');
```
This looks fishy.. are the notification drawer, notification header, and toast list **supposed** to be 3 different angular *applications*?

himdel commented on app/views/static/notification_drawer/notification-footer.html:
```diff
@@ -0,0 +1,8 @@
+     ng-class="{'disabled': !notificationGroup.notifications || notificationGroup.notifications.length === 0}"
+     ng-click="customScope.clearAllNotifications(notificationGroup)"
+    <span class="pficon pficon-close"></span>
+    <span>Clear All</span>
```
will need this file in HAML so that we can `= _("Clear All")`...

himdel commented on app/assets/javascripts/controllers/header/header_controller.js:
```diff
@@ -0,0 +1,38 @@
+
+  var refresh = function() {
+    vm.newNotifications = eventNotifications.state().unreadNotifications;
+    try {
```
Not so fond of this `$digest + catch` idiom, why is it even needed here?

Otherwise, maybe wrapping in `$timeout` is a bit more legible..?

jeff-phillips-18 commented:
Updated to add ManageIQ.notifications module.

Removed changes to miq_debug.

Put the toast list and notifications drawer under the header component so there is only one angular application.

Put the footer in a haml file to allow for translation.

Moved the update code into a timeout rather than calling $digest. This is done because the event notifications service is being updated and notifying observers outside of the notifications angular application's digest so a $apply/$digest is necessary for the components to update to the new data.
